### DISCUSSION TITLE
Erro de Float 16 e de ativação da GPU ao processar transcrição de áudio

### DISCUSSION THREAD

**Discussion by RafaelgsEspindola** [sepinf-inc/IPED#2386] [General]
Prezados, preciso de ajuda com relação ao rodar a biblioteca de transcrição de áudio deepdml/faster-whisper-large-v3-turbo-ct2 e também do whisper. Quando coloco na versão float16, apresenta o seguinte erro: "Requested float16 compute type, but the target device or backend do not support efficient float16 computation.", além disso,  quando coloco o float32, não ativa a GPU, roda apenas a CPU.  Esse mês adquirimos uma CPU para rodar o IPED, ela vem com placa de vídeo RTX 4070 SUPER, i7 14700KF e memória de 64 GB, ela veio com drive Nvidia 12.7, instalei o CUDA 12.4, cuDNN compatível com a versão 12.4 e o pytorch compatível, ademais, segui todo o processo recomendado na WIKI do IPED, porém não obtive sucesso ao rodar usando a GPU da máquina. Estou utilizando a versão 4.2 snapchat, que é uma versão muito boa e me ajudou demais, mas só preciso de ajuda com relação da GPU, pois temos várias extrações para rodar e muito deles passam de 50 mil áudios, que fica muito difícil de analisar sem uma transcrição boa.
[IPED-2024-12-18-13-23-04.log](https://github.com/user-attachments/files/18198187/IPED-2024-12-18-13-23-04.log)
[AudioTranscriptConfig (2).txt](https://github.com/user-attachments/files/18198188/AudioTranscriptConfig.2.txt)
![WhatsApp Image 2024-12-18 at 13 31 02](https://github.com/user-attachments/assets/c7b14218-e194-44ee-b55e-d6708201040d)
![WhatsApp Image 2024-12-18 at 13 36 48](https://github.com/user-attachments/assets/9cd0f388-cd75-46f6-8c73-1d59ea0cb6c6)
![WhatsApp Image 2024-12-18 at 13 35 24](https://github.com/user-attachments/assets/d820e9f1-6a7c-4a43-951b-00f032b00a95)
![WhatsApp Image 2024-12-18 at 13 44 50](https://github.com/user-attachments/assets/9173d518-ac1c-438b-ae19-f35afccaa57c)

**Comment by lfcnassif** [comment]
Hi. I'm traveling on vacation and not able to help you right now... @gfd2020 seems he is following your GPU tutorial, do you have available time to help him? Thanks in advance.

**Reply by gfd2020** [reply]
> Hi. I'm traveling on vacation and not able to help you right now... @gfd2020 seems he is following your GPU tutorial, do you have available time to help him? Thanks in advance.

Hello. I'll try to help him.

**Comment by gfd2020** [comment]
> Prezados, preciso de ajuda com relação ao rodar a biblioteca de transcrição de áudio deepdml/faster-whisper-large-v3-turbo-ct2 e também do whisper. Quando coloco na versão float16, apresenta o seguinte erro: "Requested float16 compute type, but the target device or backend do not support efficient float16 computation.", além disso, quando coloco o float32, não ativa a GPU, roda apenas a CPU. Esse mês adquirimos uma CPU para rodar o IPED, ela vem com placa de vídeo RTX 4070 SUPER, i7 14700KF e memória de 64 GB, ela veio com drive Nvidia 12.7, instalei o CUDA 12.4, cuDNN compatível com a versão 12.4 e o pytorch compatível, ademais, segui todo o processo recomendado na WIKI do IPED, porém não obtive sucesso ao rodar usando a GPU da máquina. Estou utilizando a versão 4.2 snapchat, que é uma versão muito boa e me ajudou demais, mas só preciso de ajuda com relação da GPU, pois temos várias extrações para rodar e muito deles passam de 50 mil áudios, que fica muito difícil de analisar sem uma transcrição boa. 

Oi @RafaelgsEspindola . Me tira algumas dúvidas:
1) Quando você roda com int8 a transcrição na GPU funciona?
2) No Nvidia panel está monstrando qual versão da NVCUDA64.dl ?
2) Eu também tive esses erros de não funcionar float16 e float32 mas eu pensava que era por causa de gpu antiga. A sua é bem nova. Tenta desinstalar o pytorch 12.4 e instala a 12.1 para testar ( pip install torch torchvision torchaudio --index-url https://download.pytorch.org/whl/cu121 )

4) Parece que a tabela de compatibilidade do Pytorch foi atualizada. Veja se você está atendendo todos os requisitos e versões. Pelo que eu vi o seu CUDNN está em uma versão abaixo recomendada.
![image](https://github.com/user-attachments/assets/66373980-2b0a-4de5-9d36-b71d1b4cd91f)

**Reply by RafaelgsEspindola** [reply]
Olá, @gfd2020! Muito obrigado pela ajuda e atenção. 

1. rodei com o int8 e o CUDA 11.8 e não deu certo utilizar a GPU, está aí embaixo. estou enviando também o log e audioconfig.txt
[IPED-2024-12-20-10-13-36.log](https://github.com/user-attachments/files/18211180/IPED-2024-12-20-10-13-36.log)
[AudioTranscriptConfig.txt](https://github.com/user-attachments/files/18211183/AudioTranscriptConfig.txt)
![image](https://github.com/user-attachments/assets/a5528c43-3cc6-4324-bf41-f50f2f670867)
![image](https://github.com/user-attachments/assets/f053810f-c278-474d-8354-434d0f04dae4)
![image](https://github.com/user-attachments/assets/12c72c3f-b349-4fae-bed6-987107b7f3a0)

2. A versão do NVCUDA64.dll é o 12.7, mandei também, já veio instalado quando recebemos e infelizmente, não tenho o conhecimento de alterar.

3. rodei o whisper diretamente pelo CMD da máquina e transcreveu normalmente utilizando 100% da GPU e transcreveu uma oitiva de mais de 10 min em menos de 30 seg.

Se precisar de algo a mais é só falar.

**Comment by gfd2020** [comment]
@RafaelgsEspindola , acho que está acontecendo várias coisas diferentes.
Pelo log que você me enviou, parece que está acontecendo o bug relatado aqui abaixo:
https://github.com/SYSTRAN/faster-whisper/issues/900

Esse bug é em relação ao áudio vir vazio e o whisper não consegue detectar a linguagem, então a GPU vai ficar parada mesmo pois não tem nada a transcrever pois o job falhou.

Faz um teste pra mim. Abre o zip da sua evidência e extrai uns 30 audios grandes ogg e tenta processar eles separadamente. Utilize int8 na configuração e veja se vai transcrever. 

Se funcionar, aí você tenta de novo sua evidencia inteira zipada mas espera o processo todo acabar, provavelmente a GPU vai ficar 0% enquanto não passa a parte do bug do fast-whisper ...

**Comment by RafaelgsEspindola** [comment]
@gfd2020 extraí 300 áudios em .opus, que é o formato dos áudios do Whatsapp, pelo menos a maioria, segue o resultado na primeira imagem e o primeiro log, pode verificar que continuou usando a CPU. Na segunda imagem, é o processamento com os arquivos . ogg, que você solicitou, mas que continuou utilizando a CPU. Com relação ao pytorch, estou seguindo a versão que indica no site, comecei com o CUDA 12.6, já que o drive era de 12.7, porém sem sucesso. Depois fui para o CUDA 12.4 e agora estou usando o CUDA 11.8 e continua dando erro, mesmo alterando a versão do torch e testando e reconhecendo o CUDA, tanto que rodou tranquilo novamente hoje usando o WHISPER no modo TURBO, transcreveu uma oitiva de 15 minutos em .mp4 no tempo de 1 minuto. Mandei também o arquivo de configuração. Por fim, compartilhei o arquivo do IPED que estou utilizando para verificação, caso queira. o arquivo está zipado e está no link do google drive abaixo. 
[AudioTranscriptConfig.txt](https://github.com/user-attachments/files/18214203/AudioTranscriptConfig.txt)
[IPED-2024-12-20-15-22-56.log](https://github.com/user-attachments/files/18214205/IPED-2024-12-20-15-22-56.log)
[IPED-2024-12-20-15-39-32.log](https://github.com/user-attachments/files/18214206/IPED-2024-12-20-15-39-32.log)


![image](https://github.com/user-attachments/assets/be722b09-4357-4434-8420-a69495613d67)
![image](https://github.com/user-attachments/assets/d3d401c9-6cc3-447a-bbbd-511b6d195d7d)

https://drive.google.com/drive/folders/1TbgqCTBB7lnF67ndmTAscDDfZE-7_ZPG?usp=sharing

**Comment by gfd2020** [comment]
@RafaelgsEspindola , acho que descobri o erro. Você está usando a versão snapshot 4.2 certo? Essa versão está sempre mandando para a cpu, não sei o porquê ficou escrita assim. Verifique o arquivo WhisperProcess.py que está na pasta "iped-4.2-snapshot\scripts\tasks". Provavelmente você verá perto da linha 40 o seguinte código:

```
if cudaCount > 0:
    deviceId = 'cpu'
    deviceNum = 0
else:
    deviceId = 'cpu'
    deviceNum = 0
```

O correto é o abaixo:
```
if cudaCount > 0:
    deviceId = 'cuda'
else:
    deviceId = 'cpu'
    deviceNum = 0
```


PS: Eu imagino que essa versão 4.2 snapshot não foi muito testada, aconselho cautela a utilizá-la.

**Reply by wladimirleite** [reply]
Despite this error in WhisperProcess.py, version 4.2-snapshot has been widely used here.

**Reply by lfcnassif** [reply]
Sorry for not helping here. I just checked and the code on master (last 4.2 snapshot) is correct.

**Reply by lfcnassif** [reply]
Weeks (or months) ago I helped an user and sent a similar code change to him to force CPU, since his GPU didn't have enough memory. Maybe he is wrongly distributing the patch I did for his specific case to others arbitrarily.

**Reply by RafaelgsEspindola** [reply]
Poderia disponibilizar o link desse último 4.2 snapshot que está correto? O que eu uso baixei do site ipediar, que é o único que encontrei disponibilizando essa versão e não tenho muita familiaridade com o Github para pesquisar nas discussões do site aqui, pois meus conhecimentos são medianos em TI e o debate aqui é em alto nível. Além disso, precisamos realmente usar o modelo whisper para rodar transcrição, visto que cada uma das nossas extrações passam de mais de 50 mil áudios para analisar e só o modelo whisper pode ajudar efetivamente a nossa necessidade. Agradeço bastante a ajuda de todos nessa discussão e peço perdão pelo trabalho.

**Reply by wladimirleite** [reply]
Link for the latest snapshot:
https://github.com/sepinf-inc/IPED/actions/runs/12296839061/artifacts/2311627934

**Reply by gfd2020** [reply]
> Despite this error in WhisperProcess.py, version 4.2-snapshot has been widely used here.

I'm sorry. I thought this snapshot was just for testing features. I didn't know they used it in production.

**Reply by wladimirleite** [reply]
No problem @gfd2020! 
I just want to let you guys know that there are people using the snapshot version in real cases for at least a year, so it is not completely untested/risky. 
Ideally this should not happen, but we need/want to use some new features only available in 4.2.
And the snapshot link I shared came from the last branch merged into master, so at least everything included was reviewed.

**Reply by gfd2020** [reply]
> Link for the latest snapshot: https://github.com/sepinf-inc/IPED/actions/runs/12296839061/artifacts/2311627934

@wladimirleite , Does this version have the source code too?

**Reply by wladimirleite** [reply]
Not sure which would be an easy way to check the code from that exact version, but it is pretty much the master. There were a couple of merges after that, but very small fixes.

**Reply by wladimirleite** [reply]
Are you having any trouble with that snapshot?!

**Reply by gfd2020** [reply]
> Are you having any trouble with that snapshot?!

No. I wanted to release a custom beta version before the 4.2 release came out.

**Reply by lfcnassif** [reply]
To see all source code related to a specific snapshot, just click on the commit number in the Actions tab, then click on "Browse Files".

**Comment by gfd2020** [comment]
Just a reminder. After fixing the code to perform the transcription using the GPU, I was unable to make it work. The transcription task always crashed. I discovered that the error was that a cudnn dll (cudnn_ops_infer64_8.dll) from pytorch was not in the path (I believe it must be a bug in this version). So I had to put the folder 'python\Lib\site-packages\torch\lib' in the windows path.

**Reply by lfcnassif** [reply]
@RafaelgsEspindola, was you able to run transcription on GPU with last official snapshot (not a third party version/fork) and @gfd2020's tip above?
